### PR TITLE
Pull request from karopawil's fork

### PR DESCRIPTION
**Pull request by karopawil**
Can have a look when you have time.

### CODE REVIEW

File: disruptor/claim_strategy.h [hunk 1]
```diff
@@ -95,25 +96,33 @@ class SingleThreadedStrategy {
...
+
+  int64_t GetHighestPublishedSequence(int64_t lowerBound,
+                                      int64_t availableSequence) {
+    return availableSequence;
+  }
 
  private:
   // We do not need to use atomic values since this function is called by a
   // single publisher.
+  size_t N;
```
Comment (tvdijk) [inline disruptor/claim_strategy.h]: "N" is not very expressive, and doesn't follow the naming convention used. I would call this "buffer_size_".

File: disruptor/claim_strategy.h [hunk 2]
```diff
@@ -85,8 +85,9 @@ class SingleThreadedStrategy {
     return next_sequence;
   }
 
-  bool HasAvailableCapacity(const std::vector<Sequence*>& dependents) {
-    const int64_t wrap_point = last_claimed_sequence_ + 1L - N;
+  bool HasAvailableCapacity(const std::vector<Sequence*>& dependents,
+                            int64_t cursor, int required_capacity = 1) {
```
Comment (tvdijk) [inline disruptor/claim_strategy.h:90]: cursor is unused?

File: disruptor/claim_strategy.h [hunk 3]
```diff
@@ -95,25 +96,33 @@ class SingleThreadedStrategy {
     return true;
   }
 
-  void SynchronizePublishing(const int64_t& sequence, const Sequence& cursor,
-                             const size_t& delta) {}
+  void SynchronizePublishing(const int64_t& sequence, Sequence& cursor,
+                             const size_t& delta) {
+    cursor.set_sequence(sequence);
```
Comment (tvdijk) [inline disruptor/claim_strategy.h:103]: What prompted this change?

File: disruptor/ring_buffer.h [hunk 4]
```diff
@@ -26,18 +26,18 @@
...
 #define DISRUPTOR_RING_BUFFER_H_  // NOLINT
 
-#include <array>
 #include "utils.h"
+#include <array>
 
 namespace disruptor {
 
-constexpr size_t kDefaultRingBufferSize = 1024;
+size_t kDefaultRingBufferSize = 1024;
```
Comment (tvdijk) [inline disruptor/ring_buffer.h]: Could still be const, no?

File: disruptor/ring_buffer.h [hunk 5]
```diff
@@ -48,23 +48,33 @@ class RingBuffer {
...
 
   const T& operator[](const int64_t& sequence) const {
-    return events_[sequence & (N - 1)];
+    return events_[sequence & index_mask_];
   }
 
  private:
-  std::array<T, N> events_;
+  int index_mask_;
+  T* events_;
```
Comment (tvdijk) [inline disruptor/ring_buffer.h]: Consider using a std::vector instead of a raw pointer.

File: disruptor/sequencer.h [hunk 6]
```diff
@@ -72,28 +77,38 @@ class Sequencer {
...
+    return claim_strategy_.IncrementAndGet(cursor_, gating_sequences_, delta);
   }
 
   // Publish an event and make it visible to {@link EventProcessor}s.
   //
   // @param sequence to be published.
   void Publish(const int64_t& sequence, size_t delta = 1) {
     claim_strategy_.SynchronizePublishing(sequence, cursor_, delta);
-    const int64_t new_cursor = cursor_.IncrementAndGet(delta);
+    // const int64_t new_cursor = cursor_.IncrementAndGet(delta);
```
Comment (tvdijk) [inline disruptor/sequencer.h:100]: What prompted this change?

**Review by tvdijk — CHANGES_REQUESTED** [review]
I would love to be able to take some of these changes, but I'm concerned that these fixes are mixed up with functional changes which required interface changes. Is there any way you can split this out?

**Comment by karopawil** [comment]
Hi,

Thanks for the feedback.  Suggestions improvements are always welcome. Not even sure what is there is the final interface as I haven't looked into pipelines and diamond patterns.

1. As far as I am concerned N can be changed to buffer_size_.

2. cursor value is used in MultiThreadedStrategyEx (LMAX Disruptor). To keep the interface consistent. With MultiThreadedStrategyEx Claim strategy I get a higher throughput than MultiThreadedStrategy.

3. My experience is that atomic store is much more efficient that fetch_add. In my tests I get a much higher thruput (one knows what the next sequence should be for a slot, why do a fetch and add?). And I moved the cursor.set_sequence(sequence) to SynchronizePublishing because MultiThreadedStrategyEx is intrusive and changes the cursor in IncrementAndGet. So the cursor has been forwarded. The user in this case has to check

// Only required for claim strategy MultiThreadedStrategyEx as it changes
    // the cursor.
    available_seq = s.GetHighestPublishedSequence(next_seq, available_seq);

You can have  a look at LMAX Multiproducersequencer.

4. -constexpr size_t kDefaultRingBufferSize = 1024;
+size_t kDefaultRingBufferSize = 1024;
Could still be const, no?

Yes.

5. Consider using a std::vector instead of a raw pointer.

Can be done. A matter of taste as it is a static array and should not have any effect on the performance as long as the vector is created with the required size.

6.
// Publish an event and make it visible to {@link EventProcessor}s.
   //
   // @param sequence to be published.
   void Publish(const int64_t& sequence, size_t delta = 1) {
     claim_strategy_.SynchronizePublishing(sequence, cursor_, delta);
-    const int64_t new_cursor = cursor_.IncrementAndGet(delta);
+    // const int64_t new_cursor = cursor_.IncrementAndGet(delta);
What prompted this change?

Explained above. You could change it back to IncrementAndGet and compare the throughput. example.cc can be used for that purpose.

**Comment by fsaintjacques** [comment]
It'll help if we could fix travis, I'm not even sure this compile :)

**Comment by karopawil** [comment]
I can compile this with gcc 5.4 and VS2015. What needs to be done for travis?

**Comment by karopawil** [comment]
I'll fix up my fork travis build. have to learn about it first.

**Comment by karopawil** [comment]
gcc 5.0 travis build works. clang 3.8 fails. No idea.

$ cd ${TRAVIS_BUILD_DIR} && tools/test
-- The CXX compiler identification is Clang 3.8.0
-- Check for working CXX compiler: /usr/bin/clang++-3.8
-- Check for working CXX compiler: /usr/bin/clang++-3.8 -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Boost version: 1.54.0
-- Found the following Boost libraries:
--   unit_test_framework
Coveralls: Using alternate CMake script dir: /home/travis/build/karopawil/disruptor--/tools/coveralls-cmake/cmake
COVERALLS UPLOAD: ON
-- Configuring done
-- Generating done
-- Build files have been written to: /home/travis/build/karopawil/disruptor--/build
Scanning dependencies of target claim_strategy_test_bin
[ 11%] Building CXX object CMakeFiles/claim_strategy_test_bin.dir/test/claim_strategy_test.cc.o
In file included from /home/travis/build/karopawil/disruptor--/test/claim_strategy_test.cc:29:
In file included from /usr/include/boost/test/unit_test.hpp:19:
In file included from /usr/include/boost/test/test_tools.hpp:19:
In file included from /usr/include/boost/test/predicate_result.hpp:20:
In file included from /usr/include/boost/test/utils/wrap_stringstream.hpp:26:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/4.8/../../../../include/c++/4.8/sstream:38:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/4.8/../../../../include/c++/4.8/istream:38:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/4.8/../../../../include/c++/4.8/ios:40:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/4.8/../../../../include/c++/4.8/bits/char_traits.h:39:
/usr/bin/../lib/gcc/x86_64-linux-gnu/4.8/../../../../include/c++/4.8/bits/stl_algobase.h:739:11: error: 
      overload resolution selected deleted operator '='

**Comment by karopawil** [comment]
Think gcc boost is getting used with clang. Anyway I have had enough for today.

**Comment by karopawil** [comment]
Got a working travis script for clang 3.6, 3.8 and gcc 5, 6.

**Comment by karopawil** [comment]
There are just too many changes to the interface. Not compatible.

**Pull request closed without merging**